File: src/bots/BotManager.ts
import { BybitClient } from '../exchange/BybitClient';
import { getDb } from '../persistence/db';
import { childLogger } from '../core/logger';
import type { MarketSnapshot, Regime } from '../market/MarketDataService';
import type { PortfolioState } from '../risk/RiskManager';

const log = childLogger({ module: 'bot-manager' });

// Bybit enforces a per-symbol minimum grid investment; we attempt only above
// this floor and let validateSpotGrid be the final gate. Nothing is scaled
// down artificially — the agent tries with whatever the balance allows.
const MIN_GRID_INVESTMENT_USDT = 10;

export class BotManager {
  // is_paper tags rows by environment (testnet vs mainnet) for reporting only.
  // Execution is always real against whichever exchange the client points at —
  // on testnet that's the Bybit testnet grid-bot API, not a local simulation.
  private readonly isTestnet: boolean;

  constructor(
    private readonly client: BybitClient,
    isTestnet: boolean,
  ) {
    this.isTestnet = isTestnet;
  }

  async tick(snapshots: MarketSnapshot[], portfolio: PortfolioState): Promise<void> {
    await this.monitorActiveBots();
    await this.considerNewBots(snapshots, portfolio);
  }

  private async monitorActiveBots(): Promise<void> {
    const sql = getDb();
    const bots = await sql`SELECT * FROM bot_instances WHERE status = 'active' AND is_paper = ${this.isTestnet}`;

    for (const bot of bots) {
      try {
        let detail: Record<string, unknown> | null = null;
        if (bot.bot_type === 'spot_grid' && bot.exchange_bot_id) {
          detail = await this.client.getSpotGridDetail(bot.exchange_bot_id) as Record<string, unknown>;
        } else if (bot.bot_type === 'futures_grid' && bot.exchange_bot_id) {
          detail = { bot_id: bot.exchange_bot_id };
        }

        if (detail) {
          const pnl = parseFloat(String(detail?.profit ?? detail?.realized_pnl ?? '0'));
          await sql`
            INSERT INTO bot_performance (bot_instance_id, realized_pnl, unrealized_pnl, roi)
            VALUES (${bot.id}, ${pnl}, 0, 0)
          `;
        }
      } catch (e) {
        log.warn({ e, botId: bot.id }, 'Error monitoring bot');
      }
    }
  }

  private async considerNewBots(snapshots: MarketSnapshot[], portfolio: PortfolioState): Promise<void> {
    for (const snap of snapshots) {
      const regime = classifyRegime(snap);

      // Only launch grid bots in ranging markets
      if (regime !== 'ranging') continue;

      const sql = getDb();
      const existing = await sql`
        SELECT id FROM bot_instances
        WHERE symbol = ${snap.symbol} AND status = 'active'
        LIMIT 1
      `;
      if (existing.length > 0) continue;

      const investment = portfolio.equity * 0.05; // 5% equity per bot
      if (investment < MIN_GRID_INVESTMENT_USDT) continue;

      try {
        await this.createGridBot(snap, investment);
      } catch (e) {
        log.warn({ e, symbol: snap.symbol }, 'Failed to create grid bot');
      }
    }
  }

  private async createGridBot(snap: MarketSnapshot, investment: number): Promise<void> {
    const sql = getDb();
    const atr = snap.indicators.atr14;
    const price = snap.lastPrice;
    const minPrice = Math.floor((price - atr * 3) * 100) / 100;
    const maxPrice = Math.ceil((price + atr * 3) * 100) / 100;
    const cellNumber = 10;

    const config = {
      symbol: snap.symbol, minPrice, maxPrice,
      cellNumber, totalInvestment: investment,
    };

    // Validate first (skill rule), then create on the real exchange the client
    // is pointed at (testnet or mainnet). is_paper tags the row by environment.
    await this.client.validateSpotGrid({
      symbol: snap.symbol,
      min_price: String(minPrice),
      max_price: String(maxPrice),
      cell_number: cellNumber,
      total_investment: String(investment),
    });

    const result = await this.client.createSpotGrid({
      symbol: snap.symbol,
      min_price: String(minPrice),
      max_price: String(maxPrice),
      cell_number: cellNumber,
      total_investment: String(investment),
    });

    await sql`
      INSERT INTO bot_instances (bot_type, exchange_bot_id, symbol, category, status, config, state, is_paper)
      VALUES ('spot_grid', ${result.grid_id}, ${snap.symbol}, 'spot', 'active',
        ${JSON.stringify(config)}::jsonb, '{}'::jsonb, ${this.isTestnet})
    `;
    log.info({ symbol: snap.symbol, gridId: result.grid_id, testnet: this.isTestnet }, 'Created spot grid bot');
  }

  async stopAll(): Promise<void> {
    const sql = getDb();
    const bots = await sql`SELECT * FROM bot_instances WHERE status = 'active' AND is_paper = ${this.isTestnet}`;
    for (const bot of bots) {
      try {
        if (bot.bot_type === 'spot_grid') {
          await this.client.closeSpotGrid(bot.exchange_bot_id, 1);
        } else if (bot.bot_type === 'futures_grid') {
          await this.client.closeFuturesGrid(bot.exchange_bot_id);
        }
        await sql`UPDATE bot_instances SET status = 'stopped', updated_at = now() WHERE id = ${bot.id}`;
      } catch (e) {
        log.error({ e, botId: bot.id }, 'Error stopping bot');
      }
    }
  }
}

function classifyRegime(snap: MarketSnapshot): Regime {
  const { adxValue, atrPct } = snap.indicators;
  if (atrPct > 0.05 || Math.abs(snap.fundingRate) > 0.002) return 'crisis';
  if (adxValue > 25) return 'trending';
  if (atrPct > 0.025) return 'high_volatility';
  return 'ranging';
}

File: bybit_agent/bots/bot_manager.py
"""Grid bot manager — ports src/bots/BotManager.ts.

Monitors active bots, launches new spot-grid bots in ranging regimes.

Investment sizing:
  - Per-bot allocation = max(MIN_GRID_USDT_FLOOR, equity × GRID_EQUITY_PCT)
  - Capped by MAX_CONCURRENT_GRID_BOTS so total grid exposure ≤ equity × GRID_EQUITY_PCT × max_bots
  - With $78 equity: max($15, $19.50) = $19.50 per bot, up to 3 bots = $58.50 max

Symbol source:
  - AgentLoop discovery finds affordable ranging-regime symbols every 5 min
  - BotManager takes those snapshots, filters for ranging regime, and opens grids
  - Grid range = price ± 4 × ATR14 (wider than 3× to give the bot breathing room)
"""
from __future__ import annotations

import json
import math
from typing import Any

from bybit_agent.core.logger import get_logger
from bybit_agent.market.market_data import MarketSnapshot, classify_regime
from bybit_agent.persistence.db import NeonHttpClient
from bybit_agent.risk.risk_manager import PortfolioState

log = get_logger().bind(module="bot-manager")

MIN_GRID_USDT_FLOOR = 15.0       # never allocate less than this per grid bot
GRID_EQUITY_PCT     = 0.25       # target 25% of equity per bot (was 5% — too small for small accounts)
MAX_CONCURRENT_GRID_BOTS = 3     # cap total active grids so equity isn't fully consumed


class BotManager:
    def __init__(self, client: Any, db: NeonHttpClient, is_testnet: bool) -> None:
        self._client = client
        self._db = db
        self._is_testnet = is_testnet

    async def tick(self, snapshots: list[MarketSnapshot], portfolio: PortfolioState) -> None:
        await self._monitor_active_bots()
        await self._consider_new_bots(snapshots, portfolio)

    async def stop_all(self) -> None:
        bots = await self._db.fetch(
            "SELECT * FROM bot_instances WHERE status = 'active' AND is_paper = $1",
            self._is_testnet,
        )
        for bot in bots:
            try:
                if bot["bot_type"] == "spot_grid":
                    await self._client.close_spot_grid(bot["exchange_bot_id"], 1)
                await self._db.execute(
                    "UPDATE bot_instances SET status = 'stopped' WHERE id = $1",
                    bot["id"],
                )
            except Exception as e:
                log.error("Error stopping bot", bot_id=bot["id"], error=str(e))

    async def _monitor_active_bots(self) -> None:
        bots = await self._db.fetch(
            "SELECT * FROM bot_instances WHERE status = 'active' AND is_paper = $1",
            self._is_testnet,
        )
        for bot in bots:
            try:
                detail: dict | None = None
                if bot["bot_type"] == "spot_grid" and bot.get("exchange_bot_id"):
                    detail = await self._client.get_spot_grid_detail(bot["exchange_bot_id"])
                elif bot["bot_type"] == "futures_grid" and bot.get("exchange_bot_id"):
                    detail = {"bot_id": bot["exchange_bot_id"]}

                if detail:
                    pnl = float(detail.get("profit") or detail.get("realized_pnl") or 0)
                    await self._db.execute(
                        "INSERT INTO bot_performance (bot_instance_id, realized_pnl, unrealized_pnl, roi) VALUES ($1, $2, 0, 0)",
                        bot["id"], pnl,
                    )
            except Exception as e:
                log.warning("Error monitoring bot", bot_id=bot["id"], error=str(e))

    async def _consider_new_bots(
        self, snapshots: list[MarketSnapshot], portfolio: PortfolioState
    ) -> None:
        # How many grids are already running?
        active_rows = await self._db.fetch(
            "SELECT COUNT(*) AS n FROM bot_instances WHERE status = 'active' AND is_paper = $1",
            self._is_testnet,
        )
        active_count = int((active_rows[0].get("n") or 0) if active_rows else 0)
        if active_count >= MAX_CONCURRENT_GRID_BOTS:
            log.debug("Max grid bots reached — not creating more", active=active_count)
            return

        # Per-bot investment: 25% of equity, but at least MIN_GRID_USDT_FLOOR.
        investment = max(MIN_GRID_USDT_FLOOR, portfolio.equity * GRID_EQUITY_PCT)
        # Safety cap: never put more than 50% of equity into a single grid.
        investment = min(investment, portfolio.equity * 0.50)

        if investment < MIN_GRID_USDT_FLOOR:
            log.debug("Equity too small for grid bot", equity=portfolio.equity,
                      needed=MIN_GRID_USDT_FLOOR)
            return

        for snap in snapshots:
            if active_count >= MAX_CONCURRENT_GRID_BOTS:
                break
            if classify_regime(snap) != "ranging":
                continue

            # Already have an active grid on this symbol?
            existing = await self._db.fetch(
                "SELECT id FROM bot_instances WHERE symbol = $1 AND status = 'active' LIMIT 1",
                snap.symbol,
            )
            if existing:
                continue

            try:
                await self._create_grid_bot(snap, investment)
                active_count += 1
            except Exception as e:
                log.warning("Failed to create grid bot", symbol=snap.symbol, error=str(e))

    async def _create_grid_bot(self, snap: MarketSnapshot, investment: float) -> None:
        price = snap.lastPrice
        atr = snap.indicators.atr14

        # 4× ATR range gives the grid plenty of room on small accounts.
        # floor/ceil to 2 decimal places so Bybit accepts the price levels.
        min_price = math.floor((price - atr * 4) * 100) / 100
        max_price = math.ceil((price + atr * 4) * 100) / 100
        cell_number = 10

        # Strip the perpetual suffix if present (BTCUSDT-PERP → BTCUSDT for spot).
        spot_symbol = snap.symbol.replace("-PERP", "").replace("PERP", "")

        config = {
            "symbol": spot_symbol,
            "minPrice": min_price, "maxPrice": max_price,
            "cellNumber": cell_number, "totalInvestment": investment,
        }

        log.info("Creating spot grid bot", symbol=spot_symbol,
                 price=price, min_price=min_price, max_price=max_price,
                 cell_number=cell_number, investment=round(investment, 2))

        # Validate first — Bybit enforces per-symbol minimums and price tick rules.
        await self._client.validate_spot_grid({
            "symbol": spot_symbol,
            "min_price": str(min_price), "max_price": str(max_price),
            "cell_number": cell_number, "total_investment": str(investment),
        })
        result = await self._client.create_spot_grid({
            "symbol": spot_symbol,
            "min_price": str(min_price), "max_price": str(max_price),
            "cell_number": cell_number, "total_investment": str(investment),
        })

        # Bybit returns orderId on mainnet; testnet may return 0/None.
        grid_id = (result.get("orderId") or result.get("grid_id")
                   or result.get("gridId") or result.get("botId"))

        await self._db.execute(
            """INSERT INTO bot_instances
                 (bot_type, exchange_bot_id, symbol, category, status, config, state, is_paper)
               VALUES ('spot_grid', $1, $2, 'spot', 'active', $3::jsonb, '{}'::jsonb, $4)""",
            str(grid_id) if grid_id else None, spot_symbol, json.dumps(config), self._is_testnet,
        )
        log.info("Spot grid bot created", symbol=spot_symbol,
                 grid_id=grid_id, investment=round(investment, 2),
                 testnet=self._is_testnet)
